skyd/api.go

```diff
@@ -68,44 +103,66 @@ func NewAPI(nginxHost string, nginxPort int, skydHost, skydPassword string, skyd
 	}, nil
 }
 
-// BlockSkylinks will perform an API call to skyd to block the given skylinks
-func (api *api) BlockHashes(hashes []string) error {
-	// Build the call to skyd.
-	reqBody := skyapi.SkynetBlocklistPOST{
-		Add:    hashes,
+// BlockHashes will perform an API call to skyd to block the given hashes. It
+// returns which hashes were blocked, which hashes were invalid and potentially
+// an error.
+func (api *api) BlockHashes(hashes []database.Hash) ([]database.Hash, []database.Hash, error) {
+	api.staticLogger.Debugf("blocking %v hashes", len(hashes))
+
+	// convert the hashes to strings
+	adds := make([]string, len(hashes))
+	for h, hash := range hashes {
+		adds[h] = hash.String()
+	}
+
+	// build the call to skyd.
+	reqBody, err := json.Marshal(skyapi.SkynetBlocklistPOST{
+		Add:    adds,
 		Remove: nil,
 		IsHash: true,
-	}
-	reqBodyBytes, err := json.Marshal(reqBody)
+	})
 	if err != nil {
-		return errors.AddContext(err, "failed to build request body")
+		return nil, nil, errors.AddContext(err, "failed to build request body")
 	}
 
+	// execute the request
 	url := fmt.Sprintf("http://%s:%d/skynet/blocklist?timeout=%s", api.staticNginxHost, api.staticNginxPort, skydTimeout)
-
-	api.staticLogger.Debugf("blockSkylinks: POST on %+s", url)
-	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(reqBodyBytes))
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(reqBody))
 	if err != nil {
-		return errors.AddContext(err, "failed to build request to skyd")
+		return nil, nil, errors.AddContext(err, "failed to build request to skyd")
 	}
 	req.Header.Set("User-Agent", "Sia-Agent")
 	req.Header.Set("Authorization", api.staticAuthHeader())
-
-	api.staticLogger.Debugf("blockSkylinks: headers: %+v", req.Header)
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
-		return errors.AddContext(err, "failed to make request to skyd")
+		return nil, nil, errors.AddContext(err, "failed to make request to skyd")
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
-		respBody, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			api.staticLogger.Warn(errors.AddContext(err, "failed to parse response body after a failed call to skyd").Error())
-			respBody = []byte{}
-		}
-		return errors.New(fmt.Sprintf("call to skyd failed with status '%s' and response '%s'", resp.Status, string(respBody)))
+
+	// read the response body
+	respBody, err := ioutil.ReadAll(resp.Body)
```

Comment:
Same as before. Defer also flushing the body and then use a `Decoder` instead

Reply:
done

Reply:
this is because `skyd/api.go` will be replaced by the client, I'm planning to resolve this [comment](https://github.com/SkynetLabs/blocker/pull/19#discussion_r819443292) in an issue for that